b9lab-content/3-main-concepts/10-multistore-keepers.md

```diff
@@ -108,27 +108,357 @@ The AnteHandler is theoretically optional, but still a very important component
+    Deadline: ctx.BlockTime().Add(types.MaxTurnDurationInSeconds).UTC().Format(types.DeadlineLayout),
+}
+```
+Update its value after a move:
+
+```go
+storedGame.Deadline = ctx.BlockTime().Add(types.MaxTurnDurationInSeconds).UTC().Format(types.DeadlineLayout)
+```
+Extract and verify its value when necessary:
+
+```go
+deadline, err = time.Parse(DeadlineLayout, storedGame.Deadline)
+if err != nil {
+    panic(err)
+}
+if deadline.Before(ctx.BlockTime()) {
+    // TODO
+}
+```
+
+## How To Expire Games
+
+Ok, but when do you verify that the game has expired? An interesting feature of an ABCI application is that you can do some action at the end of each block. Should you load all games and filter for those that have expired? No. That would be extremely expensive. Better keep a FIFO where fresh games are pushed back to the tail and therefore the head contains the next games to expire.
+
+In the context of the Cosmos SDK, you need to keep track of where the FIFO starts and stops by saving the corresponding game ids:
+
+```go
+const (
+    NoFifoIdKey = "-1"
+)
+type NextGame struct {
+    ...
+    FifoHead string
+    FifoTail string
+}
+```
+And to have each game know its relative position, and also the number of moves done, to assist the refunding logic on games with 0, 1 or 2+ moves:
+
+```go
+type StoredGame struct {
+    ...
+    MoveCount uint64
+    BeforeId  string
+    AfterId   string
+}
+```
+Next, and this is left as an exercise, you need to code a regular FIFO whereby:
 
-## Long-running exercise
+* Games are sent to the back when created or played on.
+* Games are removed from the FIFO when they are finished or time out.
 
-So far we have described what we wanted to do:
+## When To Expire Games
 
-* Define a game id. How is defined? Probably incrementally.
-* Save many `FullGame`s in storage for retrieval by id.
-* Save `FullGame` ids in an ordered list of their timeouts, for use in `EndBlock`.
-* Handle `CreateGameTx`, `MoveTx`, and `RejectTx`.
+Since you want to expire the games that have timed out at the end of a block, you need to hook your keeper to the right call. When building the whole application, the Cosmos SDK will call at various points into each module. The function it calls at each block's end looks like this:
 
-We need to add the other functions:
+```go
+func (am AppModule) EndBlock(ctx sdk.Context, _ abci.RequestEndBlock) []abci.ValidatorUpdate {
+    // TODO
+    return []abci.ValidatorUpdate{}
+}
+```
+This is where you write the necessary code, preferably in the keeper. For instance with:
+
+```go
+am.keeper.ForfeitExpiredGames(sdk.WrapSDKContext(ctx))
+```
+Those among you with a well-placed paranoia must be asking whether you can ensure that the execution of this `EndBlock` does not become prohibitively expensive. After all, the number of games to potentially expire is unbounded, which is a recipe for disaster in the blockchain world. Is there a situation or an attack vector that would make this is a possibility?
```

Comment:
```suggestion
Those among you with a well-placed paranoia must be asking whether you can ensure that the execution of this `EndBlock` does not become prohibitively expensive. After all, the number of games to potentially expire is unbounded, which is a recipe for disaster in the blockchain world. Is there a situation or an attack vector that would make this a possibility? And what can we do to prevent it?
```